src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContract.cs

```diff
@@ -657,55 +658,67 @@ internal static int GetId(RuntimeTypeHandle typeHandle)
             // check whether a corresponding update is required in ClassDataContract.IsNonAttributedTypeValidForSerialization
             private static DataContract CreateDataContract(int id, RuntimeTypeHandle typeHandle, Type type)
             {
-                lock (s_createDataContractLock)
+                DataContract dataContract = s_dataContractCache[id];
```

Comment:
It looks like s_dataContractCache might be read concurrently with being written, as the lock isn't taken on all accesses to s_datacontractCache, only when writing to it and when the initial check fails.  That is not a safe thing to do with a ```Dictionary<...>```, which is not thread-safe for a read concurrently with a write.

The comment on this PR indicates this is a port of a change from the full framework.  Is the full framework using a ```Hashtable``` or a ```Dictionary<...>```?  Hashtable is actually safe for reading concurrently with a single write.